rig/routing_table/ordered_covering.py

```diff
@@ -0,0 +1,635 @@
+        """
+        # Generate a new copy of the routing table
+        routing_entries = list(self.routing_table)
+        aliases = dict(aliases)
+
+        # Add the new entry
+        new_entry = RoutingTableEntry(
+            route=routing_entries[next(iter(self.entries))].route,
+            key=self.key, mask=self.mask
+        )
+        routing_entries.insert(self.insertion_index, new_entry)
+        aliases[(self.key, self.mask)] = our_aliases = set([])
+
+        # Remove all the old entries.
+        # NOTE: This relies on the fact that merging can only ever increase
+        # generality, so entry that was just inserted MUST be below the entries
+        # that we need to remove and hence doesn't affect their indices.
+        offset = 0
+        for i in sorted(self.entries):
+            entry = routing_entries.pop(i - offset)
+            offset += 1
+            km = (entry.key, entry.mask)
+
+            # Add to the alias list
+            our_aliases.update(aliases.pop(km, set([km])))
+
+        return routing_entries, aliases
+
+
+def refine_merge(merge, aliases, min_goodness):
+    """Remove entries from a merge to generate a valid merge which may be
+    applied to the routing table.
+
+    Parameters
+    ----------
+    merge : :py:class:`~.Merge`
+        Initial merge to refine.
+    aliases : {(key, mask): {(key, mask), ...}, ...}
+        Map of key-mask pairs to the sets of key-mask pairs that they actually
+        represent.
+    min_goodness : int
+        Reject merges which are worse than the minimum goodness.
+
+    Returns
+    -------
+    :py:class:`~.Merge`
+        Valid merge which may be applied to the routing table.
+    """
+    # Perform the down-check
+    merge = refine_downcheck(merge, aliases, min_goodness)
+
+    # If the merge is still sufficiently good then continue to refine it.
+    if merge.goodness > min_goodness:
+        # Perform the up-check
+        merge, changed = refine_upcheck(merge, min_goodness)
+
+        if changed and merge.goodness > min_goodness:
+            # If the up-check removed any entries we need to re-perform the
+            # down-check; but we do not need to re-perform the up-check as the
+            # down check can only move the resultant merge nearer the top of
+            # the routing table.
+            merge = refine_downcheck(merge, aliases, min_goodness)
+
+    return merge
+
+
+def overlap(key_a, mask_a, key_b, mask_b):
+    return (key_a & mask_b) == (key_b & mask_a)
+
+
+def refine_upcheck(merge, min_goodness):
+    """Remove from the merge any entries which would be covered by entries
+    between their current position and the merge insertion position.
+
+    For example, the third entry of::
+
+        0011 -> N
+        0100 -> N
+        1000 -> N
+        X000 -> NE
+
+    Cannot be merged with the first two entries because that would move it
+    below the entry with the key-mask pair of ``X000``, which would cover it.
```

Comment:
Being hyper-picky it would be worth explicitly stating that the merge would give you a key `XXXX` or something.